lib/iris/_data_manager.py

```diff
@@ -0,0 +1,392 @@
+            self._assert_axioms()
+
+        return self._real_array
+
+    @data.setter
+    def data(self, data):
+        """
+        XXX
+
+        """
+        # Ensure we have numpy-like data.
+        if not (hasattr(data, 'shape') and hasattr(data, 'dtype')):
+            data = np.asanyarray(data)
+
+        # Determine whether the __init__ has completed.
+        init_done = (self._lazy_array is not None or
+                     self._real_array is not None)
+
+        if init_done and self.shape != data.shape:
+            # The _ONLY_ data reshape permitted is converting a 0-dimensional
+            # array i.e. self.shape == () into a 1-dimensional array of length
+            # one i.e. data.shape == (1,)
+            if self.shape or data.shape != (1,):
+                emsg = 'Require data with shape {!r}, got {!r}.'
+                raise ValueError(emsg.format(self.shape, data.shape))
+
+        # Set lazy or real data, and reset the other.
+        if is_lazy_data(data):
+            self._lazy_array = data
+            self._real_array = None
+        else:
+            if not ma.isMaskedArray(data):
+                # Coerce input data to ndarray (including ndarray subclasses).
+                data = np.asarray(data)
+            self._lazy_array = None
+            self._real_array = data
+
+        # Always reset the realised dtype, as the managed data has changed.
+        self._realised_dtype = None
+
+        # Check the manager contract, as the managed data has changed.
+        self._assert_axioms()
+
+    @property
+    def dtype(self):
+        """
+        The dtype of the realised lazy data or the dtype of the real data.
+
+        """
+        if self._realised_dtype is not None:
+            result = self._realised_dtype
+        else:
+            result = self.core_data.dtype
+        return result
+
+    @property
+    def ndim(self):
+        """
+        The number of dimensions covered by the data being managed.
+
+        """
+        return len(self.shape)
+
+    @property
+    def shape(self):
+        """
+        The shape of the data being managed.
+
+        """
+        return self.core_data.shape
+
+    def copy(self, data=None, realised_dtype=None):
+        """
+        Returns a deep copy of this :class:`DataManager`.
+
+        Kwargs:
+
+        * data:
+            Replace the data of the copy with this data.
+
+        * realised_dtype:
+            Replace the intended dtype of the lazy data
+            in the copy with this :class:`~numpy.dtype`.
+
+        Returns:
+            A copy :class:`DataManager`.
+
+        """
+        memo = {}
+        return self._deepcopy(memo, data=data, realised_dtype=realised_dtype)
+
+    def has_lazy_data(self):
+        """
+        Determine whether lazy data is being managed.
+
+        Returns:
+            Boolean.
+
+        """
+        return self._lazy_array is not None
```

Comment:
I'm a fan of `has_real_data`:

* as a reader of code it is more clear what's being checked than `not has_lazy_data`, which is more than twice as hard to process as you've got to work out the whole statement from the combination of the two.
* as a user of Iris it's nice to be able to check exactly what I want rather than the inverse of what I don't want to check.
* Lines of code are cheap! We shouldn't be withholding convenience functions for the sake of fewer lines of source code...